File: Problem1.java
// ## Problem1 Find Judge (https://leetcode.com/problems/find-the-town-judge/)

class Solution {
    // Use an in-degree array to keep track of the trusts
    // Judge will be the one with (n-1) in-degree as everyone will trust it
    // Time O(V+E)
    // Space O(N) or O(V)
    public int findJudge(int n, int[][] trust) {
        int[] inDegrees = new int[n];
        for(int[] edge:trust){
            // update for outgoing edge
            inDegrees[edge[0]-1]--;

            // update for incoming edge
            inDegrees[edge[1]-1]++;
        }

        for(int i=0;i<n;i++){
            if(inDegrees[i]==(n-1)){
                return i+1;
            }
        }
        // no judge
        return -1;
    }
}

File: Problem_1.py
'''
Time Complexity - O(V+E)
Space Complexity - O(n) for maintaining the indegrees

Works on Leetcode.
'''

class Solution:
    def findJudge(self, n: int, trust: List[List[int]]) -> int:
        indegrees = [0 for i in range(n+1)] #creat indegrees array
        for tr in trust: #for each edge in the edges
            indegrees[tr[0]]-=1 #decrease indegree of the outgoing node
            indegrees[tr[1]]+=1 #increase indegree of the incoming node
        for i in range(1,n+1):
            if indegrees[i] == n-1: #the node which has the trust of everyone will have indegree no. of nodes - 1
                return i #return the node number
        return -1 #no node found return -1
        